// edgedb/dist/__esm/reflection/util/util.js
export { cardinalityUtil } from "./cardinalityUtil";
import * as genutil_1 from "./genutil";
export { genutil_1 as genutil };
export var util;
(function (util) {
    function assertNever(arg, error) {
        throw error !== null && error !== void 0 ? error : new Error(`${arg} is supposed to be of "never" type`);
    }
    util.assertNever = assertNever;
    util.deduplicate = (args) => [...new Set(args)];
    util.getFromArrayMap = (map, id) => {
        return map[id] || [];
    };
    util.defineProperty = (obj, name, def) => {
        return Object.defineProperty(obj, name, def);
    };
    util.defineGetter = (obj, name, getter) => {
        return Object.defineProperty(obj, name, {
            get: getter,
            enumerable: true,
        });
    };
    util.defineMethod = (obj, name, method) => {
        obj[name] = method.bind(obj);
        return obj;
    };
    function flatMap(array, callbackfn) {
        return Array.prototype.concat(...array.map(callbackfn));
    }
    util.flatMap = flatMap;
    function omitDollarPrefixed(object) {
        const obj = {};
        for (const key of Object.keys(object)) {
            if (!key.startsWith("$")) {
                obj[key] = object[key];
            }
        }
        return obj;
    }
    util.omitDollarPrefixed = omitDollarPrefixed;
})(util || (util = {}));
