// edgedb/dist/__esm/errors/index.js
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { EdgeDBError } from "./base";
import * as tags from "./tags";
export { EdgeDBError } from "./base";
export * from "./tags";
export class InternalServerError extends EdgeDBError {
    get code() {
        return 16777216;
    }
}
export class UnsupportedFeatureError extends EdgeDBError {
    get code() {
        return 33554432;
    }
}
export class ProtocolError extends EdgeDBError {
    get code() {
        return 50331648;
    }
}
export class BinaryProtocolError extends ProtocolError {
    get code() {
        return 50397184;
    }
}
export class UnsupportedProtocolVersionError extends BinaryProtocolError {
    get code() {
        return 50397185;
    }
}
export class TypeSpecNotFoundError extends BinaryProtocolError {
    get code() {
        return 50397186;
    }
}
export class UnexpectedMessageError extends BinaryProtocolError {
    get code() {
        return 50397187;
    }
}
export class InputDataError extends ProtocolError {
    get code() {
        return 50462720;
    }
}
export class ParameterTypeMismatchError extends InputDataError {
    get code() {
        return 50462976;
    }
}
export class StateMismatchError extends InputDataError {
    get code() {
        return 50463232;
    }
}
export class ResultCardinalityMismatchError extends ProtocolError {
    get code() {
        return 50528256;
    }
}
export class CapabilityError extends ProtocolError {
    get code() {
        return 50593792;
    }
}
export class UnsupportedCapabilityError extends CapabilityError {
    get code() {
        return 50594048;
    }
}
export class DisabledCapabilityError extends CapabilityError {
    get code() {
        return 50594304;
    }
}
export class QueryError extends EdgeDBError {
    get code() {
        return 67108864;
    }
}
export class InvalidSyntaxError extends QueryError {
    get code() {
        return 67174400;
    }
}
export class EdgeQLSyntaxError extends InvalidSyntaxError {
    get code() {
        return 67174656;
    }
}
export class SchemaSyntaxError extends InvalidSyntaxError {
    get code() {
        return 67174912;
    }
}
export class GraphQLSyntaxError extends InvalidSyntaxError {
    get code() {
        return 67175168;
    }
}
export class InvalidTypeError extends QueryError {
    get code() {
        return 67239936;
    }
}
export class InvalidTargetError extends InvalidTypeError {
    get code() {
        return 67240192;
    }
}
export class InvalidLinkTargetError extends InvalidTargetError {
    get code() {
        return 67240193;
    }
}
export class InvalidPropertyTargetError extends InvalidTargetError {
    get code() {
        return 67240194;
    }
}
export class InvalidReferenceError extends QueryError {
    get code() {
        return 67305472;
    }
}
export class UnknownModuleError extends InvalidReferenceError {
    get code() {
        return 67305473;
    }
}
export class UnknownLinkError extends InvalidReferenceError {
    get code() {
        return 67305474;
    }
}
export class UnknownPropertyError extends InvalidReferenceError {
    get code() {
        return 67305475;
    }
}
export class UnknownUserError extends InvalidReferenceError {
    get code() {
        return 67305476;
    }
}
export class UnknownDatabaseError extends InvalidReferenceError {
    get code() {
        return 67305477;
    }
}
export class UnknownParameterError extends InvalidReferenceError {
    get code() {
        return 67305478;
    }
}
export class SchemaError extends QueryError {
    get code() {
        return 67371008;
    }
}
export class SchemaDefinitionError extends QueryError {
    get code() {
        return 67436544;
    }
}
export class InvalidDefinitionError extends SchemaDefinitionError {
    get code() {
        return 67436800;
    }
}
export class InvalidModuleDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436801;
    }
}
export class InvalidLinkDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436802;
    }
}
export class InvalidPropertyDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436803;
    }
}
export class InvalidUserDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436804;
    }
}
export class InvalidDatabaseDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436805;
    }
}
export class InvalidOperatorDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436806;
    }
}
export class InvalidAliasDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436807;
    }
}
export class InvalidFunctionDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436808;
    }
}
export class InvalidConstraintDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436809;
    }
}
export class InvalidCastDefinitionError extends InvalidDefinitionError {
    get code() {
        return 67436810;
    }
}
export class DuplicateDefinitionError extends SchemaDefinitionError {
    get code() {
        return 67437056;
    }
}
export class DuplicateModuleDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437057;
    }
}
export class DuplicateLinkDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437058;
    }
}
export class DuplicatePropertyDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437059;
    }
}
export class DuplicateUserDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437060;
    }
}
export class DuplicateDatabaseDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437061;
    }
}
export class DuplicateOperatorDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437062;
    }
}
export class DuplicateViewDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437063;
    }
}
export class DuplicateFunctionDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437064;
    }
}
export class DuplicateConstraintDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437065;
    }
}
export class DuplicateCastDefinitionError extends DuplicateDefinitionError {
    get code() {
        return 67437066;
    }
}
export class SessionTimeoutError extends QueryError {
    get code() {
        return 67502080;
    }
}
export class IdleSessionTimeoutError extends SessionTimeoutError {
    get code() {
        return 67502336;
    }
}
export class QueryTimeoutError extends SessionTimeoutError {
    get code() {
        return 67502592;
    }
}
export class TransactionTimeoutError extends SessionTimeoutError {
    get code() {
        return 67504640;
    }
}
export class IdleTransactionTimeoutError extends TransactionTimeoutError {
    get code() {
        return 67504641;
    }
}
export class ExecutionError extends EdgeDBError {
    get code() {
        return 83886080;
    }
}
export class InvalidValueError extends ExecutionError {
    get code() {
        return 83951616;
    }
}
export class DivisionByZeroError extends InvalidValueError {
    get code() {
        return 83951617;
    }
}
export class NumericOutOfRangeError extends InvalidValueError {
    get code() {
        return 83951618;
    }
}
export class AccessPolicyError extends InvalidValueError {
    get code() {
        return 83951619;
    }
}
export class IntegrityError extends ExecutionError {
    get code() {
        return 84017152;
    }
}
export class ConstraintViolationError extends IntegrityError {
    get code() {
        return 84017153;
    }
}
export class CardinalityViolationError extends IntegrityError {
    get code() {
        return 84017154;
    }
}
export class MissingRequiredError extends IntegrityError {
    get code() {
        return 84017155;
    }
}
export class TransactionError extends ExecutionError {
    get code() {
        return 84082688;
    }
}
export class TransactionConflictError extends TransactionError {
    get code() {
        return 84082944;
    }
}
Object.defineProperty(TransactionConflictError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: { [tags.SHOULD_RETRY]: true }
});
export class TransactionSerializationError extends TransactionConflictError {
    get code() {
        return 84082945;
    }
}
Object.defineProperty(TransactionSerializationError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: { [tags.SHOULD_RETRY]: true }
});
export class TransactionDeadlockError extends TransactionConflictError {
    get code() {
        return 84082946;
    }
}
Object.defineProperty(TransactionDeadlockError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: { [tags.SHOULD_RETRY]: true }
});
export class ConfigurationError extends EdgeDBError {
    get code() {
        return 100663296;
    }
}
export class AccessError extends EdgeDBError {
    get code() {
        return 117440512;
    }
}
export class AuthenticationError extends AccessError {
    get code() {
        return 117506048;
    }
}
export class AvailabilityError extends EdgeDBError {
    get code() {
        return 134217728;
    }
}
export class BackendUnavailableError extends AvailabilityError {
    get code() {
        return 134217729;
    }
}
Object.defineProperty(BackendUnavailableError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: { [tags.SHOULD_RETRY]: true }
});
export class BackendError extends EdgeDBError {
    get code() {
        return 150994944;
    }
}
export class UnsupportedBackendFeatureError extends BackendError {
    get code() {
        return 150995200;
    }
}
export class LogMessage extends EdgeDBError {
    get code() {
        return 4026531840;
    }
}
export class WarningMessage extends LogMessage {
    get code() {
        return 4026597376;
    }
}
export class ClientError extends EdgeDBError {
    get code() {
        return 4278190080;
    }
}
export class ClientConnectionError extends ClientError {
    get code() {
        return 4278255616;
    }
}
export class ClientConnectionFailedError extends ClientConnectionError {
    get code() {
        return 4278255872;
    }
}
export class ClientConnectionFailedTemporarilyError extends ClientConnectionFailedError {
    get code() {
        return 4278255873;
    }
}
Object.defineProperty(ClientConnectionFailedTemporarilyError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        [tags.SHOULD_RECONNECT]: true,
        [tags.SHOULD_RETRY]: true,
    }
});
export class ClientConnectionTimeoutError extends ClientConnectionError {
    get code() {
        return 4278256128;
    }
}
Object.defineProperty(ClientConnectionTimeoutError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        [tags.SHOULD_RECONNECT]: true,
        [tags.SHOULD_RETRY]: true,
    }
});
export class ClientConnectionClosedError extends ClientConnectionError {
    get code() {
        return 4278256384;
    }
}
Object.defineProperty(ClientConnectionClosedError, "tags", {
    enumerable: true,
    configurable: true,
    writable: true,
    value: {
        [tags.SHOULD_RECONNECT]: true,
        [tags.SHOULD_RETRY]: true,
    }
});
export class InterfaceError extends ClientError {
    get code() {
        return 4278321152;
    }
}
export class QueryArgumentError extends InterfaceError {
    get code() {
        return 4278321408;
    }
}
export class MissingArgumentError extends QueryArgumentError {
    get code() {
        return 4278321409;
    }
}
export class UnknownArgumentError extends QueryArgumentError {
    get code() {
        return 4278321410;
    }
}
export class InvalidArgumentError extends QueryArgumentError {
    get code() {
        return 4278321411;
    }
}
export class NoDataError extends ClientError {
    get code() {
        return 4278386688;
    }
}
export class InternalClientError extends ClientError {
    get code() {
        return 4278452224;
    }
}
