// edgedb/dist/__esm/datatypes/range.js
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export class Range {
    constructor(_lower, _upper, _incLower = true, _incUpper = false) {
        Object.defineProperty(this, "_lower", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _lower
        });
        Object.defineProperty(this, "_upper", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _upper
        });
        Object.defineProperty(this, "_incLower", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _incLower
        });
        Object.defineProperty(this, "_incUpper", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: _incUpper
        });
        Object.defineProperty(this, "_isEmpty", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: false
        });
    }
    get lower() {
        return this._lower;
    }
    get upper() {
        return this._upper;
    }
    get incLower() {
        return this._incLower;
    }
    get incUpper() {
        return this._incUpper;
    }
    get isEmpty() {
        return this._isEmpty;
    }
    static empty() {
        const range = new Range(null, null);
        range._isEmpty = true;
        return range;
    }
    toJSON() {
        return this.isEmpty
            ? { empty: true }
            : {
                lower: this._lower,
                upper: this._upper,
                inc_lower: this._incLower,
                inc_upper: this._incUpper,
            };
    }
}
