// edgedb/dist/__esm/options.js
import * as errors from "./errors";
export function defaultBackoff(attempt) {
    return 2 ** attempt * 100 + Math.random() * 100;
}
export var IsolationLevel;
(function (IsolationLevel) {
    IsolationLevel["Serializable"] = "SERIALIZABLE";
})(IsolationLevel || (IsolationLevel = {}));
export var RetryCondition;
(function (RetryCondition) {
    RetryCondition[RetryCondition["TransactionConflict"] = 0] = "TransactionConflict";
    RetryCondition[RetryCondition["NetworkError"] = 1] = "NetworkError";
})(RetryCondition || (RetryCondition = {}));
class RetryRule {
    constructor(attempts, backoff) {
        Object.defineProperty(this, "attempts", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "backoff", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.attempts = attempts;
        this.backoff = backoff;
    }
}
export class RetryOptions {
    constructor(attempts = 3, backoff = defaultBackoff) {
        Object.defineProperty(this, "default", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "overrides", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.default = new RetryRule(attempts, backoff);
        this.overrides = new Map();
    }
    withRule(condition, attempts, backoff) {
        const def = this.default;
        const overrides = new Map(this.overrides);
        overrides.set(condition, new RetryRule(attempts !== null && attempts !== void 0 ? attempts : def.attempts, backoff !== null && backoff !== void 0 ? backoff : def.backoff));
        const result = Object.create(RetryOptions.prototype);
        result.default = def;
        result.overrides = overrides;
        return result;
    }
    getRuleForException(err) {
        let result;
        if (err instanceof errors.TransactionConflictError) {
            result = this.overrides.get(RetryCondition.TransactionConflict);
        }
        else if (err instanceof errors.ClientError) {
            result = this.overrides.get(RetryCondition.NetworkError);
        }
        return result !== null && result !== void 0 ? result : this.default;
    }
    static defaults() {
        return new RetryOptions();
    }
}
export class TransactionOptions {
    constructor({ isolation = IsolationLevel.Serializable, readonly = false, deferrable = false, } = {}) {
        Object.defineProperty(this, "isolation", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "readonly", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "deferrable", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.isolation = isolation;
        this.readonly = readonly;
        this.deferrable = deferrable;
    }
    static defaults() {
        return new TransactionOptions();
    }
}
export class Session {
    constructor({ module = "default", moduleAliases = {}, config = {}, globals = {}, } = {}) {
        Object.defineProperty(this, "module", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "moduleAliases", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "config", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "globals", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.module = module;
        this.moduleAliases = moduleAliases;
        this.config = config;
        this.globals = globals;
    }
    withModuleAliases({ module, ...aliases }) {
        return new Session({
            ...this,
            module: module !== null && module !== void 0 ? module : this.module,
            moduleAliases: { ...this.moduleAliases, ...aliases },
        });
    }
    withConfig(config) {
        return new Session({
            ...this,
            config: { ...this.config, ...config },
        });
    }
    withGlobals(globals) {
        return new Session({
            ...this,
            globals: { ...this.globals, ...globals },
        });
    }
    _serialise() {
        const state = {};
        if (this.module !== "default") {
            state.module = this.module;
        }
        const _aliases = Object.entries(this.moduleAliases);
        if (_aliases.length) {
            state.aliases = _aliases;
        }
        if (Object.keys(this.config).length) {
            state.config = this.config;
        }
        const _globals = Object.entries(this.globals);
        if (_globals.length) {
            state.globals = _globals.reduce((globals, [key, val]) => {
                globals[key.includes("::") ? key : `${this.module}::${key}`] = val;
                return globals;
            }, {});
        }
        return state;
    }
    static defaults() {
        return defaultSession;
    }
}
const defaultSession = new Session();
export class Options {
    constructor({ retryOptions = RetryOptions.defaults(), transactionOptions = TransactionOptions.defaults(), session = Session.defaults(), } = {}) {
        Object.defineProperty(this, "retryOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "transactionOptions", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "session", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.retryOptions = retryOptions;
        this.transactionOptions = transactionOptions;
        this.session = session;
    }
    withTransactionOptions(opt) {
        return new Options({
            ...this,
            transactionOptions: opt instanceof TransactionOptions ? opt : new TransactionOptions(opt),
        });
    }
    withRetryOptions(opt) {
        return new Options({
            ...this,
            retryOptions: opt instanceof RetryOptions
                ? opt
                : new RetryOptions(opt.attempts, opt.backoff),
        });
    }
    withSession(opt) {
        return new Options({
            ...this,
            session: opt,
        });
    }
    static defaults() {
        return new Options();
    }
}
