// edgedb/dist/__esm/primitives/chars.js
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2019-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
export const $0 = ord("0");
export const $1 = ord("1");
export const $A = ord("A");
export const $C = ord("C");
export const $D = ord("D");
export const $E = ord("E");
export const $H = ord("H");
export const $I = ord("I");
export const $K = ord("K");
export const $L = ord("L");
export const $M = ord("M");
export const $O = ord("O");
export const $P = ord("P");
export const $Q = ord("Q");
export const $R = ord("R");
export const $S = ord("S");
export const $T = ord("T");
export const $V = ord("V");
export const $X = ord("X");
export const $Y = ord("Y");
export const $Z = ord("Z");
export const $b = ord("b");
export const $j = ord("j");
export const $m = ord("m");
export const $n = ord("n");
export const $o = ord("o");
export const $p = ord("p");
export const $r = ord("r");
export const $s = ord("s");
export const $v = ord("v");
export function ord(str) {
    const ch = str.charCodeAt(0);
    if (ch <= 0 || ch >= 255) {
        throw new TypeError(`char "${ch}" is outside ASCII`);
    }
    return ch & 0xff;
}
export function chr(ch) {
    if (ch <= 0 || ch >= 255) {
        throw new TypeError(`char "${ch}" is outside ASCII`);
    }
    return String.fromCharCode(ch);
}
