// edgedb/dist/__esm/primitives/queues.js
/*!
 * This source file is part of the EdgeDB open source project.
 *
 * Copyright 2020-present MagicStack Inc. and the EdgeDB authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { InternalClientError } from "../errors";
export class LifoQueue {
    constructor() {
        Object.defineProperty(this, "_promises", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_resolvers", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "_rejecters", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this._resolvers = [];
        this._rejecters = [];
        this._promises = [];
    }
    _add() {
        this._promises.push(new Promise((resolve, reject) => {
            this._resolvers.push(resolve);
            this._rejecters.push(reject);
        }));
    }
    push(item) {
        if (!this._resolvers.length) {
            this._add();
        }
        const resolve = this._resolvers.shift();
        this._rejecters.shift();
        if (!resolve) {
            throw new InternalClientError("resolve function was null or undefined when attempting to push.");
        }
        resolve(item);
    }
    get() {
        if (!this._promises.length) {
            this._add();
        }
        const promise = this._promises.pop();
        if (!promise) {
            throw new InternalClientError("promise was null or undefined when attempting to get.");
        }
        return promise;
    }
    cancelAllPending(err) {
        const rejecters = this._rejecters;
        this._rejecters = [];
        this._resolvers = [];
        for (const reject of rejecters) {
            reject(err);
        }
    }
    get length() {
        return this._promises.length - this._resolvers.length;
    }
    get pending() {
        return Math.max(0, this._resolvers.length - this._promises.length);
    }
}
